src/foam/dao/JdbcDAO.java

```diff
@@ -0,0 +1,384 @@
+/**
+ * @license
+ * Copyright 2017 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.dao;
+
+import foam.core.*;
+import foam.dao.pg.IndexedPreparedStatement;
+import foam.mlang.order.Comparator;
+
+import foam.mlang.predicate.Predicate;
+
+import javax.sql.DataSource;
+import java.lang.Exception;
+import java.sql.*;
+
+import java.sql.SQLException;
+import java.util.*;
+
+/*
+  We assume that the database is created by a script when the system starts the first time.
+  When the system restarts at any point in time it should verify that the database is already created otherwise it should create it.
+  When creating the database, we only do a CREATE SCHEMA sql instruction, then we create the app user and password and grant him full privileges on this database.
+  Any other database objects (tables) will be created on the fly by the application (when methods of this class are called).
+
+ */
+public class JdbcDAO extends AbstractDAO{
+
+
+  // Holds the relevant properties (column names) of the table
+  protected List<PropertyInfo> properties_;
+
+  private ThreadLocal<StringBuilder> threadLocalBuilder_ = new ThreadLocal<StringBuilder>(){
+
+    @Override
+    protected StringBuilder initialValue() {
+      return new StringBuilder();
+    }
+
+    @Override
+    public StringBuilder get() {
+      StringBuilder builder = super.get();
+      builder.setLength(0);
+      return builder;
+    }
+
+  };
+
+  protected String tableName_;
+
+  // Holds a reference to the connection pool ( .getConnection() )
+  protected static DataSource dataSource_;
+
+  public JdbcDAO(X x, ClassInfo of) throws java.sql.SQLException, ClassNotFoundException {
+    setX(x);
+    setOf(of);
+
+    // Get the system global dataSource with its system global pool
+    dataSource_ = JDBCPooledDataSource.getDataSource(x);
+
+    tableName_ = of.getObjClass().getSimpleName().toLowerCase();
+
+    getObjectProperties(of);
+
+    if ( ! createTable(of) ) {
+      // Table already created (may happen after a system restart).
+    }
+
+  }
+
+  @Override
+  public FObject put_(X x, FObject obj) {
+    Connection c = null;
+    IndexedPreparedStatement stmt = null;
+    ResultSet resultSet = null;
+
+    try {
+      c = dataSource_.getConnection();
```

Comment:
Should get dataSource and connection from Context 'X', rather than using statics. Static values restrict composability. What happens if we want to connect to multiple data-sources in the same system? With the context I can register multiple instances, but using static variables precludes me from doing this.